idx/metric_definition.go

```diff
@@ -0,0 +1,552 @@
+	copy(b, []byte(mn.String()))
+	return nil
+}
+
+// UnmarshalBinary is required to use custom marshaling as an extension
+// in msgp
+func (mn *MetricName) UnmarshalBinary(b []byte) error {
+	err := mn.setMetricName(string(b))
+	return err
+}
+
+//msgp:shim TagKeyValue as:string using:TagKeyValue.createTagKeyValue/parseTagKeyValue
+
+// TagKeyValue stores a Key/Value pair. The strings
+// are interned in an object store before they are assigned.
+type TagKeyValue struct {
+	Key   uintptr
+	Value uintptr
+}
+
+// String returns a Key/Value pair in the form of
+// 'key=value'
+func (t *TagKeyValue) String() string {
+	var bld strings.Builder
+	return t.string(&bld)
+}
+
+func (t *TagKeyValue) string(bld *strings.Builder) string {
+	key, err := IdxIntern.GetStringFromPtr(t.Key)
+	if err != nil {
+		panic(fmt.Sprintf("idx: Failed to retrieve interned tag key: %v", err))
+	}
+	val, err := IdxIntern.GetStringFromPtr(t.Value)
+	if err != nil {
+		panic(fmt.Sprintf("idx: Failed to retrieve interned tag value: %v", err))
+	}
+
+	bld.WriteString(key)
+	bld.WriteString("=")
+	bld.WriteString(val)
+
+	return bld.String()
+}
+
+func (t TagKeyValue) createTagKeyValue() string {
+	bld := strings.Builder{}
+
+	key, err := IdxIntern.GetStringFromPtr(t.Key)
+	if err != nil {
+		log.Error("idx: Failed to retrieve interned tag key: ", err)
+		internError.Inc()
+	}
+	val, err := IdxIntern.GetStringFromPtr(t.Value)
+	if err != nil {
+		log.Error("idx: Failed to retrieve interned tag value: ", err)
+		internError.Inc()
+	}
+
+	bld.WriteString(key)
+	bld.WriteString("=")
+	bld.WriteString(val)
+
+	return bld.String()
+}
+
+func parseTagKeyValue(tag string) TagKeyValue {
+	var tkv TagKeyValue
+	if strings.Contains(tag, ";") {
+		log.Errorf("idx: Tag %q has an invalid format, ignoring", tag)
+		invalidTag.Inc()
+		return TagKeyValue{}
+	}
+	eqPos := strings.Index(tag, "=")
+	if eqPos < 0 {
+		log.Errorf("idx: Tag %q has an invalid format, ignoring", tag)
+		invalidTag.Inc()
+		return TagKeyValue{}
+	}
+	key, err := IdxIntern.AddOrGet([]byte(tag[:eqPos]), true)
+	if err != nil {
+		log.Errorf("idx: Failed to intern tag %q, %v", tag, err)
+		internError.Inc()
+		return TagKeyValue{}
+	}
+	tkv.Key = key
+
+	value, err := IdxIntern.AddOrGet([]byte(tag[eqPos+1:]), true)
+	if err != nil {
+		log.Errorf("idx: Failed to intern tag %q, %v", tag, err)
+		internError.Inc()
+		return TagKeyValue{}
+	}
+	tkv.Value = value
+	return tkv
+}
+
+type KeyValuesSlice []TagKeyValue
+
+func (kvs KeyValuesSlice) Len() int      { return len(kvs) }
+func (kvs KeyValuesSlice) Swap(i, j int) { kvs[i], kvs[j] = kvs[j], kvs[i] }
+func (kvs KeyValuesSlice) Less(i, j int) bool {
+	k1, _ := IdxIntern.GetStringFromPtr(kvs[i].Key)
+	k2, _ := IdxIntern.GetStringFromPtr(kvs[j].Key)
+	return k1 < k2
+}
+
+// TagKeyValues stores a slice of all of the Tag Key/Value pair combinations for a MetricDefinition
+type TagKeyValues struct {
+	KeyValues KeyValuesSlice
```

Comment:
why does this need a struct to encapsulate the `.KeyValues` property? couldn't the `TagKeyValues` methods just be methods of `KeyValuesSlice`?

Reply:
I don't recall off the top of my head, but there was a reason behind doing it that way. I'm pretty sure it goes back to msgp issues I was having. I'll look into it, but I think it needs to stay like this.